compiler/qsc_rir/src/rir.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc_data_structures::index_map::IndexMap;
+
+/// The root of the RIR.
+pub struct Program {
+    pub blocks: IndexMap<BlockId, Block>,
+    pub callables: IndexMap<CallableId, Callable>,
+    pub entry: CallableId,
+}
+
+/// A unique identifier for a block in a RIR program.
+#[derive(Clone, Copy, Default)]
+pub struct BlockId(u32);
+
+/// A block is a collection of instructions.
+pub struct Block(pub Vec<Instruction>);
+
+/// A unique identifier for a callable in a RIR program.
+#[derive(Clone, Copy, Default)]
+pub struct CallableId(u32);
+
+/// A callable.
+pub struct Callable {
+    /// The ID of the callable.
+    pub id: CallableId,
+    /// The name of the callable.
+    pub name: String,
+    /// The input type of the callable.
+    pub input_type: Vec<Ty>,
+    /// The output type of the callable.
+    pub output_type: Option<Ty>,
+    /// The callable body.
+    /// N.B. `None` bodys represent an intrinsic.
+    pub body: Option<BlockId>,
+}
+
+pub enum Instruction {
```

Comment:
Thinking about it some more, I think removing the concept of statement means some instructions need to have a Variable field for their result. Take LogicalNot as an example. Where does the result of that instruction get written to? I think Call will need an Option<Variable> and others that always produce a value will need Variable as a field.

Reply:
Good catch! I'll update the instructions that need a variable field.

Reply:
Updated.